tetragonpod/controller/tetragonpod_controller.go

```diff
@@ -0,0 +1,194 @@
+	if exists {
+		// CASE 2.1: TetragonPod already exists, update it.
+		l.Info("Pod Updated: ", "Name", req.Name, "Namespace", req.Namespace)
+		updateTetragonPod(ctx, r, l, pod, tetragonPod)
+		return ctrl.Result{}, nil
+	}
+
+	// CASE 2.2: TetragonPod does not exist, create new.
+	l.Info("Pod Created: ", "Name", req.Name, "Namespace", req.Namespace)
+	createTetragonPod(ctx, pod, r, l)
+	return ctrl.Result{}, nil
+}
+
+// deleteTetragonPod deletes the corresponding tetragonPod resource.
+func deleteTetragonPod(ctx context.Context, r *TetragonPodReconciler, req ctrl.Request, l logr.Logger) (ctrl.Result, error) {
+	l.Info("Deleting the corresponding tetragonPod resource")
+
+	// create the pod to be deleted
+	tetragonPodToDelete := &ciliumiov1alpha1.TetragonPod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+
+	// Check if corresponding tetragonPod resource exists
+	if err := r.Get(ctx, req.NamespacedName, tetragonPodToDelete); err != nil {
+		l.Error(err, "Corresponding tetragonPod Resource does not exist")
+		return ctrl.Result{}, err
+	}
+
+	// Delete the Pod
+	if err := r.Delete(ctx, tetragonPodToDelete); err != nil {
+		l.Error(err, "Failed to delete the tetragonPod resource")
+		return ctrl.Result{}, err
+	}
+
+	// Pod deleted successfully.
+	l.Info("TetragonPod deleted successfully")
+	return ctrl.Result{}, nil
+}
+
+// updateTetragonPod updates the corresponding tetragonPod resource.
+func updateTetragonPod(ctx context.Context, r *TetragonPodReconciler, l logr.Logger, pod *corev1.Pod, tetragonPod *ciliumiov1alpha1.TetragonPod) (ctrl.Result, error) {
+	l.Info("Updating the corresponding tetragonPod resource")
+
+	// get the new IP address of the Pod.
+	newIP := ciliumiov1alpha1.PodIP{IP: pod.Status.PodIP}
+	tetragonPod.Status.PodIP = newIP.IP
+	// add the new IP to the beginning of the array.
+	tetragonPod.Status.PodIPs = append([]ciliumiov1alpha1.PodIP{newIP}, tetragonPod.Status.PodIPs...)
+
+	// Update the tetragonPod resource.
+	if err := r.Update(ctx, tetragonPod); err != nil {
+		l.Error(err, "Failed to update the tetragonPod resource")
+		return ctrl.Result{}, err
+	}
+	l.Info("TetragonPod updated successfully")
+	return ctrl.Result{}, nil
+}
+
+// checkIfAlreadyExists checks if the corresponding tetragonPod resource already exists.
+func checkIfAlreadyExists(ctx context.Context, req ctrl.Request, r *TetragonPodReconciler, l logr.Logger) (*ciliumiov1alpha1.TetragonPod, bool) {
+
+	// check if the tetragonPod already exists for that pod.
+	tetragonPod, err := getTetragonPod(ctx, req, r, l)
+	if err != nil {
+		return nil, false
+	}
+	return tetragonPod, true
+}
+
+// createTetragonPod creates a tetragonPod resource.
+func createTetragonPod(ctx context.Context, pod *corev1.Pod, r *TetragonPodReconciler, l logr.Logger) (ctrl.Result, error) {
+	// create the tetragon pod since the pod exists
+	if pod.Status.Phase == corev1.PodRunning {
```

Comment:
instead of waiting for the pod to start, we want to create `TetragonPod` as soon as we have all the necessary info.

Reply:
I added that to reduce the logs being generated, because the controller is triggered for every state change too. I should use `log.Debug` instead of `log.Info` as you suggested. Should I instead for the `pod.Status.Phase == corev1.PodPending` ?

Reply:
Just re-confirming. The controller will wait for all the necessary information it needs, once done, it will create the respective tetragonPods right ?